File: android/src/main/kotlin/com/syncrhrohealth/bandpass_ble_connection/core/command/DeviceNameSetCmd.kt
package com.syncrhrohealth.bandpass_ble_connection.core.command

import com.syncrhrohealth.bandpass_ble_connection.core.command.builder.CommandBuilder
import com.syncrhrohealth.bandpass_ble_connection.core.enum.CommandCode
import com.syncrhrohealth.bandpass_ble_connection.core.handler.DeviceHandler

/** Data format
 *
 *  char[20]       name
 *
 * */

object DeviceNameSetCmd {
    fun send(name: String, handler: DeviceHandler) {
        val nameData = nameTo20Bytes(name)
        val builder = CommandBuilder.build(CommandCode.PACKET_CMD_DEVICE_NAME_SET, nameData)

        handler.write(builder)
    }

    private fun nameTo20Bytes(name: String): ByteArray {
        val raw = name.toByteArray(Charsets.US_ASCII)
        return when {
            raw.size == 20 -> raw
            raw.size < 20  -> {
                val padded = ByteArray(20)
                System.arraycopy(raw, 0, padded, 0, raw.size)
                padded
            }
            else -> raw.copyOf(20)
        }
    }
}

File: ios/Classes/Core/Command/DeviceNameSetCmd.swift
//
//  DeviceInfoNameSetCmd.swift
//  Runner
//
//  Created by MAC on 11/3/25.
//

import Foundation

/// Data format:
/// - `char[20] name`
struct DeviceNameSetCmd {
    static func send(name: String, handler: DeviceHandler) {
        let nameData = nameTo20Bytes(name: name)
        
        // Convert [UInt8] to Data using the correct function from Data+ext.swift
        let nameDataAsData = Data.byteArrayToData(nameData)
        
        let packet = CommandBuilder.build(commandCode: .packetCmdDeviceNameSet, data: nameDataAsData)
        handler.write(data: packet)
    }

    private static func nameTo20Bytes(name: String) -> [UInt8] {
        let raw = Array(name.prefix(20).utf8)  // Convert string to UTF-8 byte array (ASCII compatible)
        
        if raw.count == 20 {
            return raw
        } else {
            return raw + Array(repeating: 0, count: 20 - raw.count) // Pad with zeros
        }
    }
}
